### DISCUSSION TITLE
Hundreds of `visible: false` layers. Slow performance.

### DISCUSSION THREAD

**Discussion by juneidy** [visgl/deck.gl#7021] [Q&A] [answered]
Hi there, I have noticed a performance degradation in a map where there is 645 layers but only 12 is visible.

Profiling doesn't show anything useful other than "Function call" that took more than 100ms of nothing that explains the CPU cycle.

GPU utilisation seems to be very minimum. CPU however is cycling for something I cannot understand.

![image](https://user-images.githubusercontent.com/13762553/173013606-10abc6b1-6a55-4f26-ae1e-535076399cef.png)

Most of my layers are tile layers using GeoJsonLayer as sub-layers (so I understand that creates heaps of sub layers).

I understand that [deck.gl performance optimization](https://deck.gl/docs/developer-guide/performance#number-of-layers) says that it's only tested to be performing good within 100 layers. However, I did not expect that even invisible layers could affect the performance so significantly.

Also FYI: the profiling is simply doing a pan in a very close vicinity (vectors already loaded in GPU)

**Comment by Pessimistress** [comment, marked as answer]
You gave little information on how you implemented your TileLayer. Are you using React? Are you rerendering the layer on every view state change?

If so, this may be relevant: https://github.com/visgl/deck.gl/issues/6810

**Reply by juneidy** [reply]
I am not using React. Everything is vanilla javascript.

I am not rerendering the layer on every view state change, or at least I do not think so.

I do not think the deck lifecycle is affecting the slow performance. I believe the profiling will show signs of deck code if that is the case.

Edit:
> You gave little information on how you implemented your TileLayer

What sort of information do you need?

Edit2:
More info: The same set of visible layers perform brilliantly when the `visible: false` layers are removed.

**Reply by juneidy** [reply]
![image](https://user-images.githubusercontent.com/13762553/173030459-3162a9e4-c8eb-4a9e-8cdb-8fcf2c5843e8.png)

Another screenshot of the full profiling. As you can see, there is little to no javascript code that took CPU cycle.

**Reply by Pessimistress** [reply]
As a start, what version of deck.gl are you using, and how many TileLayers are you rendering?

**Reply by juneidy** [reply]
Deck version 8.7.11
Total of 643 TileLayer with only 12 being visible.

Also, in mind that in my code, all invisible tile vector layer has empty `data`.

The `data` is fetched asynchronously when tile is visible.

**Reply by Pessimistress** [reply]
Depending on your screen size, each TileLayer can generate dozens of sub layers, so we could be talking about over 10K layers here.

I would say deck can comfortably handle a few hundreds layers, but this is too many. With `visible: false` each layer still holds memory on both CPU and GPU.

I would recommend you use one TileLayer and have getTileData grab different data based on the visibility of your sources.

**Reply by juneidy** [reply]
```
layers.filter(l => l instanceof TileVectorLayer && !l.isVisible()).map(l => l._state.layer.state.tileset._selectedTiles).reduce((sum, arr) => sum + arr.length, 0)
15144
```

well.... in that case, do you think it's better for me to add/remove the layer to/from `layers` array in `deck.setProps({ layer })` when a layer's visibility is toggled?

Combining the layers seems like a major refactor for my code. Also needing to handle situation where I have some raster layers in between some vector layers.

**Reply by Pessimistress** [reply]
Right so you have 643 TileLayers + 15144 GeoJsonLayers + (n ScatterplotLayer, PathLayer and SolidPolygonLayer, n >= 15144). On every animation frame all of them need to be check for dirty flags, and that is a lot even without redraw.

The `visible` prop basically exchanges memory usage for faster toggling. If your app expects only a small fraction of all these layers ever being visible, removing/adding the selected layers is going to be more performant for sure.

**Reply by juneidy** [reply]
Looks like my other computer was acting weirdly on Friday. The profiling in different computer now shows that javascript code that the processing takes place.

![image](https://user-images.githubusercontent.com/13762553/173286546-e64b7a8d-099c-41cc-a7bc-5535804d08fa.png)

And yes, trimming the layers array of the invisible ones definitely made the performance much much better. I might have to rethink how many layers I'll be expecting to be visible.

![image](https://user-images.githubusercontent.com/13762553/173286731-8af08cdb-c25d-4ed5-8228-f6ae483cee85.png)
